### PR TITLE
AzureBlobDataAccessor class using Azure v12 lib

### PR DESCRIPTION
**Pull request by lightningrob**
First of 2 PR's, next one will integrate this with AzureCloudDestination and remove old libs.

### CODE REVIEW

File: ambry-cloud/src/main/java/com.github.ambry.cloud/azure/AzureBlobDataAccessor.java [hunk 1]
```diff
@@ -0,0 +1,309 @@
...
+  public boolean uploadIfNotExists(BlobId blobId, long inputLength, CloudBlobMetadata cloudBlobMetadata,
+      InputStream blobInputStream) throws BlobStorageException, IOException {
+    BlobRequestConditions blobRequestConditions = new BlobRequestConditions().setIfNoneMatch("*");
+    azureMetrics.blobUploadRequestCount.inc();
+    Timer.Context storageTimer = azureMetrics.blobUploadTime.time();
+    try {
+      BlockBlobClient blobClient = getAzureBlobReference(blobId, true);
+      cloudBlobMetadata.setCloudBlobName(getAzureBlobName(blobId));
+      Map<String, String> metadata = getMetadataMap(cloudBlobMetadata);
+      Response<BlockBlobItem> response =
```
Comment (ankagrawal) [inline ambry-cloud/src/main/java/com.github.ambry.cloud/azure/AzureBlobDataAccessor.java]: Minor: We might not need to set response here. We arent using it anywhere.

File: ambry-cloud/src/main/java/com.github.ambry.cloud/azure/AzureBlobDataAccessor.java [hunk 2]
```diff
@@ -0,0 +1,309 @@
...
+        }
+        String etag = blobProperties.getETag();
+        Map<String, String> metadata = blobProperties.getMetadata();
+        // Update only if value has changed
+        String textValue = String.valueOf(value);
+        if (!textValue.equals(metadata.get(fieldName))) {
+          metadata.put(fieldName, textValue);
+          // Set condition to ensure we don't clobber another update
+          BlobRequestConditions blobRequestConditions = new BlobRequestConditions().setIfMatch(etag);
+          Response response = blobClient.setMetadataWithResponse(metadata, blobRequestConditions, null, Context.NONE);
```
Comment (ankagrawal) [inline ambry-cloud/src/main/java/com.github.ambry.cloud/azure/AzureBlobDataAccessor.java]: same as above.

File: ambry-cloud/src/main/java/com.github.ambry.cloud/azure/AzureBlobDataAccessor.java [hunk 3]
```diff
@@ -0,0 +1,309 @@
...
+      }
+    }
+    return containerClient;
+  }
+
+  /**
+   * @return the name of the Azure storage container where blobs in the specified partition are stored.
+   * @param partitionPath the lexical path of the Ambry partition.
+   */
+  String getAzureContainerName(String partitionPath) {
```
Comment (ankagrawal) [inline ambry-cloud/src/main/java/com.github.ambry.cloud/azure/AzureBlobDataAccessor.java]: minor: can be a private method.

File: ambry-cloud/src/main/java/com.github.ambry.cloud/azure/AzureBlobDataAccessor.java [hunk 4]
```diff
@@ -0,0 +1,309 @@
...
+    String rawContainerName = clusterName + SEPARATOR + partitionPath;
+    return rawContainerName.toLowerCase();
+  }
+
+  /**
+   * Get the blob name to use in Azure Blob Storage
+   * @param blobId The {@link BlobId} to store.
+   * @return An Azure-friendly blob name.
+   */
+  String getAzureBlobName(BlobId blobId) {
```
Comment (ankagrawal) [inline ambry-cloud/src/main/java/com.github.ambry.cloud/azure/AzureBlobDataAccessor.java]: minor: can be private

File: ambry-cloud/src/main/java/com.github.ambry.cloud/azure/AzureBlobDataAccessor.java [hunk 5]
```diff
@@ -0,0 +1,306 @@
...
+      if (!knownContainers.contains(containerName)) {
+        try {
+          if (!containerClient.exists()) {
+            containerClient.create();
+            logger.info("Created container {}", containerName);
+          }
+        } catch (BlobStorageException ex) {
+          if (ex.getErrorCode() != BlobErrorCode.CONTAINER_ALREADY_EXISTS) {
+            logger.error("Failed to create container {}", containerName);
+            throw ex;
```
Comment (cgtz) [inline ambry-cloud/src/main/java/com.github.ambry.cloud/azure/AzureBlobDataAccessor.java]: I wonder if this is really an error. Doesn't it just mean that some other thread/vcr got to creating the container first while this thread was doing the existence check?

File: ambry-rest/src/test/java/com.github.ambry.rest/NettyMultipartRequestTest.java [hunk 6]
```diff
@@ -294,7 +294,7 @@ public void prepareTest() throws Exception {
       request.prepare();
       fail("Prepare should have failed because there was more than one part named Part-1");
     } catch (RestServiceException e) {
-      assertEquals("Unexpected RestServiceErrorCode", RestServiceErrorCode.MalformedRequest, e.getErrorCode());
+      assertEquals("Unexpected RestServiceErrorCode", RestServiceErrorCode.BadRequest, e.getErrorCode());
```
Comment (cgtz) [inline ambry-rest/src/test/java/com.github.ambry.rest/NettyMultipartRequestTest.java]: I will try to look into why this behavior changed after this pr is merged

File: ambry-cloud/src/main/java/com.github.ambry.cloud/azure/AzureBlobDataAccessor.java (hunk 5 above)
Comment (lightningrob) [reply ambry-cloud/src/main/java/com.github.ambry.cloud/azure/AzureBlobDataAccessor.java]: only throws exception if the error is something other than already-exists.

**Pull request merged**